dbcon/mysql/rbo_apply_rewrite_distinct.cpp

```diff
@@ -0,0 +1,181 @@
+/* Copyright (C) 2025 MariaDB Corporation
+
+   This program is free software; you can redistribute it and/or
+   modify it under the terms of the GNU General Public License
+   as published by the Free Software Foundation; version 2 of
+   the License.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+   MA 02110-1301, USA. */
+
+#include "rulebased_optimizer.h"
+
+#include "calpontselectexecutionplan.h"
+#include "aggregatecolumn.h"
+#include "simplecolumn.h"
+#include "existsfilter.h"
+#include "functioncolumn.h"
+#include "logicoperator.h"
+
+namespace optimizer
+{
+
+bool matchRewriteDistinct(execplan::CalpontSelectExecutionPlan& csep)
+{
+  return csep.distinct() && csep.tableList().size() > 0;
```

Comment:
Are you sure that we want to re-write any DISTINCT queries? Originally we discussed we use the rule for the category with DISTINCT + WFs.

Reply:
Why not? At the moment, with this rewrite, we lost only optimization of the `SELECT DISTINCT ... LIMIT n` (without `ORDER BY`) query. But we can use an approximation for an early cut-off of the aggregation phase in this case.